### PR TITLE
8319781: RISC-V: Refactor UseRVV related checks

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you review the patch to refactor the code related UseRVV checks?
Thanks!

There are some code (flag setting/checking) depending on UseRVV's value, these code should be refactored, especially after the change of https://bugs.openjdk.org/browse/JDK-8319408:
1. some code needs to get the final UseRVV's value rather than the initial value, e.g. ChaCha20 intrinsic.
2. refactored to be more readable.
3. also add note to make sure the future code get the final UseRVV value instead of inital value.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319781](https://bugs.openjdk.org/browse/JDK-8319781): RISC-V: Refactor UseRVV related checks (**Enhancement** - P4)

### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [f9337946](https://git.openjdk.org/jdk/pull/16580/files/f93379461c1bd027c24fc6b648462e2f97210ae1)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) ⚠️ Review applies to [7d05b867](https://git.openjdk.org/jdk/pull/16580/files/7d05b86763b1540ad257fbaf53e455085bbee434)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16580/head:pull/16580` \
`$ git checkout pull/16580`

Update a local copy of the PR: \
`$ git checkout pull/16580` \
`$ git pull https://git.openjdk.org/jdk.git pull/16580/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16580`

View PR using the GUI difftool: \
`$ git pr show -t 16580`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16580.diff">https://git.openjdk.org/jdk/pull/16580.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16580#issuecomment-1803577880)

### CODE REVIEW

**Comment by robehn** [comment]
Hey, how is the changes to SpecialEncodeISOArray related ?

File: src/hotspot/cpu/riscv/vm_version_riscv.cpp [hunk 1]
```diff
@@ -261,6 +251,24 @@ void VM_Version::initialize() {
...
+      // read vector length from vector CSR vlenb
+      _initial_vector_length = cpu_vector_length();
+    }
+  }
+
+#ifdef COMPILER2
+  c2_initialize();
+#endif // COMPILER2
+
+  // NOTE: Make sure codes dependent on UseRVV are put at the behind of c2_initialize(),
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/vm_version_riscv.cpp:269]:
```suggestion
  // NOTE: Make sure codes dependent on UseRVV are put after c2_initialize(),
```

**Comment by Hamlin-Li** [comment]
> Hey, how is the changes to SpecialEncodeISOArray related ?

This patch is also to respond the comment at https://github.com/openjdk/jdk/pull/16481#discussion_r1386040152

File: src/hotspot/cpu/riscv/vm_version_riscv.cpp (hunk 1 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/vm_version_riscv.cpp:269]: Thanks!

**Comment by robehn** [comment]
> > Hey, how is the changes to SpecialEncodeISOArray related ?
> 
> This patch is also to respond the comment at [#16481 (comment)](https://github.com/openjdk/jdk/pull/16481#discussion_r1386040152)

Ok, looks good!

**Review by robehn — APPROVED** [review]

File: src/hotspot/cpu/riscv/vm_version_riscv.cpp [hunk 2]
```diff
@@ -289,17 +293,9 @@ void VM_Version::c2_initialize() {
 
   if (!UseRVV) {
     FLAG_SET_DEFAULT(SpecialEncodeISOArray, false);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/vm_version_riscv.cpp:295]: Could you please remove this line while you are on it? I don't think we need to reset `SpecialEncodeISOArray` here.

**Review by RealFYang — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/vm_version_riscv.cpp (hunk 2 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/vm_version_riscv.cpp:295]: I'm not sure, as there are some code in share/classfile/vmIntrinsics.cpp which depends on SpecialEncodeISOArray's value:
```
  case vmIntrinsics::_encodeISOArray:
  case vmIntrinsics::_encodeAsciiArray:
  case vmIntrinsics::_encodeByteISOArray:
    if (!SpecialEncodeISOArray) return true;
    break;
```

Comment (RealFYang) [reply src/hotspot/cpu/riscv/vm_version_riscv.cpp:296]: The reason is that `SpecialEncodeISOArray` is a DIAGNOSTIC option which I think is for debugging / trouble-shooting purpose. The other four ones are `SpecialStringCompareTo`, `SpecialStringIndexOf`, `SpecialStringEquals` and `SpecialArraysEquals`. This code snippet checking for `SpecialEncodeISOArray` is in function `vmIntrinsics::disabled_by_jvm_flags`. So I think we are safe to have this change. In fact, `SpecialEncodeISOArray` and the others are not checked or modified by other platforms either.

Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/vm_version_riscv.cpp]: It makes sense, Thanks!

**Review by RealFYang — APPROVED** [review]
Updated change looks good. Thanks.

**Comment by Hamlin-Li** [comment]
@robehn @luhenry @RealFYang Thanks for your reviewing!

/integrate

**Pull request closed without merging**